README.md
UTIL-ARCHIVER-FCrackZip
=======================

Open Source ZIP password cracker.

LICENSE
===============
- GNU GPL V2

COMPILE TOOLS
===============
* emx + Andy Zabolotny's gcc 3.0.3 port

AUTHORS
===============
* Port: Paul Hartman

LINKS
===============
* 


NEWS
version 0.2.4
	- fixed -v mode for zip cracking and a few minor nits.

version 0.2.3
	- improved speed of assembly version ever so slightly. More unrolling
          might result in even better speed on ppro and above(!).
	- fixed a compiler compaitiblity problem with newer versions of gcc. Always
          try the --validate option with the method you want to use before you
          trust fcrackzip!
	- --help now displays the method number in addition to the name.
        - applied short options fix from Tomislav Greguric
          <greguric@stud.uni-frankfurt.de>
        - fcrackzip now detects extended header format files, treating them
          corerctly instead of silently not finding any passwords.
        - first try at a dictionary mode. not fast, but working.

version 0.2.2
	- portability fixes.
        - improved cpmask cracking by 16-20% by implementing the
          simple&effective ideas of Tim Jones <twj@cix.compulink.co.uk>.
        - applied slightly modified patch by greguric@stud.uni-frankfurt.de.

version 0.2.1
	- verbose mode for cpmask, fixed verbose mode for zipcrack.
        - default method is now respected.
        - great (but non-algorithmic) speedup for cpmask.

version 0.2.0
	- portability fixes
        - cpmask support (!).

version 0.1.3
	- portability fixes

version 0.1.2
	- fixed the --init-password option
	- added zipinfo program

version 0.1.1
	- fixed small configure bug
	- fixed --length option to accept a single number

version 0.1
	- implemented multiple methods
	- accept more types of zipfiles
	- new --length option for easier searching
	- new --method option to choose cracking method at runtime

version 0.0.5  Wed Aug 12 22:27:44 CEST 1998
	- removed USE_LARGE_MTAB. nobody has such a good memory system
	- more configuration cleanup
	- max. number of chars/password now 40
	- fixed a bug with pentium assembly (reported by Pavel Semjanov)

version 0.0.4  Sat Jul 25 03:53:12 CEST 1998
	- configure now detects more features itself
	- increased the speed on ppro by ~60%, pentium code unchanged
	- removed SYNTHETIC_MULTIPLY support. the next thing to drop
	  is VERY_LARGE_CACHE & LARGE_MULT_TAB
	- added --use-unzip option.
	- added more useful help.

version 0.0.3
	- many bugfixes
	- added primitive unzip-checking

version 0.0.2
	- now we are a bit faster than fzc (about 4%)
	  wasn't difficult, though. (this is on a non-MMX pentium!)
	- benchmarking works (more) reliably
	- some bugfixes

version 0.0.1
	- added autoconf && automake



crack.c
#include "config.h"
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include "crack.h"

u8 pw[MAX_PW + 1] = "aaaaaa";
u8 *pw_end;			/* must point to the trailing zero byte.  */

u8 files[MAX_FILES * HEADER_SIZE];
const char *file_path[MAX_FILES];
int file_count;

u8 bf_next[256];
u8 bf_last;

int verbosity;

static u8 mult_tab[16384];

static u32
fgetu32 (FILE * f)
{
  return (fgetc (f) << 0) |
    (fgetc (f) << 8) |
    (fgetc (f) << 16) |
    (fgetc (f) << 24);
}

static u32
fgetu16 (FILE * f)
{
  return (fgetc (f) << 0) |
    (fgetc (f) << 8);
}

static void
load_zip (const char *path)
{
  FILE *f = fopen (path, "rb");

  if (!f)
    {
      fprintf (stderr, "skipping '%s': %s\n", path, strerror (errno));
      goto out;
    }

  while (!feof (f))
    {
      u32 id = fgetu32 (f);

      if (id == 0x04034b50UL)
	{
	  u16 version = fgetu16 (f);
	  u16 flags = fgetu16 (f);
	  u16 compression_method = fgetu16 (f);
	  u16 lastmodtime = fgetu16 (f);
	  u16 lastmoddate = fgetu16 (f);
	  u32 crc32 = fgetu32 (f);
	  u32 compr_size = fgetu32 (f);
	  u32 uncompr_size = fgetu32 (f);
	  u16 name_len = fgetu16 (f);
	  u16 extra_field_len = fgetu16 (f);

	  char zip_path[1024];

	  /* these are unused.  */
	  (void) lastmoddate;
	  (void) lastmodtime;
	  (void) compression_method;
	  (void) version;

	  if (name_len < 1024)
	    {
	      fread (zip_path, name_len, 1, f);
	      zip_path[name_len] = 0;
	    }
	  else
	    {
	      fprintf (stderr, "filename too long (>1023 bytes), skipping zipfile\n");
	      goto out;
	    }

	  fseek (f, extra_field_len, SEEK_CUR);

	  if (flags & 1)
	    {
	      if (compr_size >= 12)
		{
		  u8 *file = files + HEADER_SIZE * file_count;
		  fread (file, FILE_SIZE, 1, f);

                  if (flags & 8)
                    {
                      /* extended header format? */
                      file[FILE_SIZE] = lastmodtime >> 8;
                      file[FILE_SIZE + 1] = lastmodtime;
                    }
                  else
                    {
                      file[FILE_SIZE] = crc32 >> 24;
                      file[FILE_SIZE + 1] = crc32 >> 16;
                    }

		  file_path[file_count] = strdup (path);

		  if (verbosity)
		    printf ("found file '%s', (size cp/uc %6lu/%6lu, flags %lx, chk %02x%02x)\n",
                            zip_path, (unsigned long) compr_size, (unsigned long) uncompr_size, (unsigned long) flags,
                            file[FILE_SIZE], file[FILE_SIZE+1]);

		  if (++file_count >= MAX_FILES)
		    {
		      if (verbosity)
			printf ("%d file maximum reached, skipping further files\n", MAX_FILES);

		      goto out;
		    }

		  compr_size -= 12;
		}
	      else
		{
		  fprintf (stderr, "'%s' is corrupted, skipping zipfile\n", zip_path);
		  goto out;
		}
	    }
	  else if (verbosity)
	    printf ("'%s' is not encrypted, skipping\n", zip_path);

	  fseek (f, compr_size, SEEK_CUR);
	}
      else if (id == 0x08074b50UL)	/* extended local sig (?)  */
	{
	  fseek (f, 12, SEEK_CUR);
	}
      else if (id == 0x30304b50UL)
	{
	  /* ignore */
	}
      else if (id == 0x02014b50UL || id == 0x06054b50UL)
	{
	  goto out;
	}
      else
	{
	  fprintf (stderr, "found id %08lx, '%s' is not a zipfile ver 2.xx, skipping\n",
		   (unsigned long) id, path);
	  goto out;
	}
    }

out:
  fclose (f);
}

#include "cpmask.c"
#include "crackdef.c"


zipinfo.c
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#ifdef HAVE_STDBOOL_H
#include <stdbool.h>
#else
typedef enum { FALSE = 0, TRUE = 1 } bool;
#endif
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#else
#include "getopt.h"
#endif

#include <string.h>
#include <errno.h>

static u32 fgetu32 (FILE *f)
{
  return (fgetc(f) <<  0) |
         (fgetc(f) <<  8) |
         (fgetc(f) << 16) |
         (fgetc(f) << 24);
}

static u32 fgetu16 (FILE *f)
{
  return (fgetc(f) <<  0) |
         (fgetc(f) <<  8);
}

static void parse_zip (char *path)
{
  FILE *f = fopen (path, "rb");
  
  if (!f)
    {
      fprintf (stderr, "skipping '%s': %s\n", path, strerror (errno));
      goto out;
    }
  
  while (!feof (f))
    {
    u32 id = fgetu32 (f);
    
    if (id == 0x04034b50UL)
      {
        u16 version		= fgetu16 (f);
        u16 flags		= fgetu16 (f);
        u16 compression_method 	= fgetu16 (f);
        u16 lastmodtime		= fgetu16 (f);
        u16 lastmoddate		= fgetu16 (f);
        u32 crc32		= fgetu32 (f);
        u32 compr_size		= fgetu32 (f);
        u32 uncompr_size	= fgetu32 (f);
        u16 name_len		= fgetu16 (f);
        u16 extra_field_len	= fgetu16 (f);
        
        char zip_path [1024];

        (void)crc32; /* suppress warning */
        
        /* these are unused.  */
        (void) lastmoddate;
        (void) lastmodtime;
        (void) compression_method;
        (void) version;
        
        if (name_len < 1024)
          {
            fread (zip_path, name_len, 1, f);
            zip_path [name_len] = 0;
          }
        else
          {
            fprintf (stderr, "filename too long (>1023 bytes), skipping zipfile\n");
            goto out;
          }
        
        fseek (f, extra_field_len, SEEK_CUR);
        
        printf ("found file '%s', size %ld (%ld)", zip_path, (long)uncompr_size, (long)compr_size);
        
        if (flags & 1)
          printf (", encrypted");
        
        if (compr_size >= 23)
          {
            u8 file[24];
            fread (file, 24, 1, f);
            
            printf ("\n%02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x",
                    file[0], file[1], file[ 2], file[ 3], 
                    file[4], file[5], file[ 6], file[ 7], 
                    file[8], file[9], file[10], file[11]);
                    
            printf (" : %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x",
                    file[12], file[13], file[14], file[15], 
                    file[16], file[17], file[18], file[19], 
                    file[20], file[21], file[22], file[23]);
                    
            compr_size -= 24;
          }
        
        printf ("\n");
        
        fseek (f, compr_size, SEEK_CUR);
      }
    else if (id == 0x08074b50UL) /* extended local sig (?)  */
      {
        fseek (f, 12, SEEK_CUR);
      }
    else if (id == 0x30304b50UL)
      {
        /* ignore */
      }
    else if (id == 0x02014b50UL || id == 0x06054b50UL)
      {
        goto out;
      }
    else
      {
        fprintf (stderr, "found id %08lx, '%s' is not a zipfile ver 2.xx, skipping\n",
                 (unsigned long)id, path);
        goto out;
      }
    }
  
  out:
  fclose (f);
}

static void usage (int ec)
{
  fprintf (stderr, "\n"
                   PACKAGE " version " VERSION ", zipinfo - tell me about a zip file\n"
                   "written by Marc Lehmann <pcg@goof.com> You can find more info on\n"
                   "http://www.goof.com/pcg/marc/\n"
                   "\n"
                   "USAGE: zipinfo file...                the zipfiles to parse\n"
                   "\n"
           );
  
  exit (ec);
}

static struct option options[] = {
  { "help"		, no_argument		, 0, 'h' },
  { 0			, 0			, 0,  0  },
};

int main (int argc, char *argv[])
{
  int option_index = 0;
  int c;
  
  while ((c = getopt_long (argc, argv, "", options, &option_index)) != -1)
    switch (c)
      {
      case 'h':
        usage (0);
        break;
      }
  
  if (optind >= argc)
    {
      fprintf (stderr, "you have to specify one or more zip files (try --help)\n");
      exit (1);
    }
  
  for (; optind < argc; optind++)
    parse_zip (argv[optind]);
  
  return 0;
}


crackdef.c
/*
 * this file is automatically generated from zipcrack.c, do NOT modify
 */

#if 0
#elif defined(__i386__) || defined(__i386) || defined(i386)
# define ARCH_i386 1
# undef METHOD
# define METHOD 1
# undef TARGET_CPU
# define TARGET_CPU 5
# undef USE_MULT_TAB
# undef init_crack_pw
# define init_crack_pw METHOD1_init_crack_pw
# undef crack_pw
# define crack_pw METHOD1_crack_pw
#  if TARGET_CPU==COMPILE_CPU && USE_MULT_TAB
#   define DEFAULT_METHOD 1
#  endif
#  include "zipcrack.c"
# undef METHOD
# define METHOD 2
# undef TARGET_CPU
# define TARGET_CPU 5
# define USE_MULT_TAB 1
# undef init_crack_pw
# define init_crack_pw METHOD2_init_crack_pw
# undef crack_pw
# define crack_pw METHOD2_crack_pw
#  if TARGET_CPU==COMPILE_CPU && USE_MULT_TAB
#   define DEFAULT_METHOD 2
#  endif
#  include "zipcrack.c"
# undef METHOD
# define METHOD 3
# undef TARGET_CPU
# define TARGET_CPU 6
# undef USE_MULT_TAB
# undef init_crack_pw
# define init_crack_pw METHOD3_init_crack_pw
# undef crack_pw
# define crack_pw METHOD3_crack_pw
#  if TARGET_CPU==COMPILE_CPU && USE_MULT_TAB
#   define DEFAULT_METHOD 3
#  endif
#  include "zipcrack.c"
# undef METHOD
# define METHOD 4
# undef TARGET_CPU
# define TARGET_CPU 6
# define USE_MULT_TAB 1
# undef init_crack_pw
# define init_crack_pw METHOD4_init_crack_pw
# undef crack_pw
# define crack_pw METHOD4_crack_pw
#  if TARGET_CPU==COMPILE_CPU && USE_MULT_TAB
#   define DEFAULT_METHOD 4
#  endif
#  include "zipcrack.c"
# undef ARCH_i386
#elif 1
# undef METHOD
# define METHOD 1
# undef USE_MULT_TAB
# undef init_crack_pw
# define init_crack_pw METHOD1_init_crack_pw
# undef crack_pw
# define crack_pw METHOD1_crack_pw
#  if USE_MULT_TAB
#   define DEFAULT_METHOD 1
#  endif
#  include "zipcrack.c"
# undef METHOD
# define METHOD 2
# define USE_MULT_TAB 1
# undef init_crack_pw
# define init_crack_pw METHOD2_init_crack_pw
# undef crack_pw
# define crack_pw METHOD2_crack_pw
#  if USE_MULT_TAB
#   define DEFAULT_METHOD 2
#  endif
#  include "zipcrack.c"
#else
#warn no architecture compiled in
#endif

#ifndef DEFAULT_METHOD
#define DEFAULT_METHOD 0
#endif

int default_method = DEFAULT_METHOD;

method methods[] = {
#if 0
#elif defined(__i386__) || defined(__i386) || defined(i386)
  { "cpmask", init_cpmask, crack_cpmask, load_img },
  { "zip1, TARGET_CPU=5", METHOD1_init_crack_pw, METHOD1_crack_pw, load_zip },
  { "zip2, TARGET_CPU=5, USE_MULT_TAB", METHOD2_init_crack_pw, METHOD2_crack_pw, load_zip },
  { "zip3, TARGET_CPU=6", METHOD3_init_crack_pw, METHOD3_crack_pw, load_zip },
  { "zip4, TARGET_CPU=6, USE_MULT_TAB", METHOD4_init_crack_pw, METHOD4_crack_pw, load_zip },
#elif 1
  { "cpmask", init_cpmask, crack_cpmask, load_img },
  { "zip1", METHOD1_init_crack_pw, METHOD1_crack_pw, load_zip },
  { "zip2, USE_MULT_TAB", METHOD2_init_crack_pw, METHOD2_crack_pw, load_zip },
#else
#endif
  { 0, 0, 0, 0 }
};



zipcrack.c
/*
 * zipcrack.c - where the 'crack' routine is located
 */

/*
 * guiding comments for gencrack.pl
 *
 * MTH cpmask init_cpmask crack_cpmask load_img
 *
 * LCL init_crack_pw crack_pw
 * LOD load_zip
 *
 * ARC ARCH_i386
 * DEF USE_MULT_TAB
 * DEF TARGET_CPU 5 $ 6
 * DFT TARGET_CPU==COMPILE_CPU && USE_MULT_TAB
 * DEP defined(__i386__) || defined(__i386) || defined(i386)
 *
 * DEF USE_MULT_TAB
 * DFT USE_MULT_TAB
 * DEP 1
 *
 * END
 */

#include "crc32.h"

#undef USE_GCC_X86_ASM
#define USE_GCC_X86_ASM (ARCH_i386 && USE_GCC_ASM && USE_MULT_TAB)

static void init_crack_pw (void)
{
#if USE_MULT_TAB
  u16 t;
  for (t = 0; t < 16384; t++)
    mult_tab[t] = ((t*4+3) * (t*4+2) >> 8) & 0xff;
#endif
}

static int crack_pw (gen_func genfunc, callback_func cbfunc)
{
  int changed = -1;
  int crack_count = 0;
  u32 key_stack[(MAX_PW+1) * 3] = { 0x12345678UL, 0x23456789UL, 0x34567890UL };
  u32 *sp;
#if USE_GCC_X86_ASM
  int dummy; /* dummy output.  */
#endif

  sp = 0; /* to calm down dumb compilers */

  do
    {
      int count = file_count;
      int count2 = 0;
      u32 key0, key1, key2;
      u8 *p;
      u8 *b = files;
      
      if (changed < 0)
        {
          changed = strlen (pw);
          pw_end = pw + changed;
          sp = key_stack + changed * 3;
        }
      
      sp -= changed * 3;
      p = (u8 *)pw_end - changed;
      
      if (++crack_count >= 1000000 && verbosity)
        {
          printf ("checking pw %-40.40s\r", pw), fflush (stdout);
          crack_count = 0;
        }
      
#  if USE_GCC_X86_ASM
      asm("
        movl (%7),%0
        movl 4(%7),%1
        movl 8(%7),%2
        xorl %3,%3
        movb (%5),%b3
1:      xorb %b0,%b3
        shrl $8,%0
        incl %5
        xorl %c6(,%3,4),%0
        movb %b0,%b3
        addl $12,%7
        movl %0,(%7)
        addl %1,%3
        imul $134775813,%3
        leal 1(%3),%1
        shrl $24,%3
        movl %1,4(%7)
        xorb %b2,%b3
        shrl $8,%2
        xorl %c6(,%3,4),%2
        movb (%5),%b3
        orb  %b3,%b3
        movl %2,8(%7)
        jnz  1b
        " : "=c" (key0), "=a" (key1), "=b" (key2), "=&d" (dummy), "=D" (sp)
          : "S" (p), "i" (crc_32_tab), "D" (sp)
      );
#  else
      key0 = *sp++;
      key1 = *sp++;
      key2 = *sp++;
      do {
        *sp++ = key0 = crc32 (key0, *p++);
        *sp++ = key1 = (key1 + (u8)key0) * 134775813 + 1;
        *sp++ = key2 = crc32 (key2, key1 >> 24);
      } while (*p);
      
      sp -= 3;
#  endif
      
      do
        {
          u8 target, pre_target;
          
#         if USE_GCC_X86_ASM
            asm("
              movl $-12,%%edi
              pushl %%ebx
              addl $12,%3
              pushl %%ecx
              xorl %%edx,%%edx
              pushl %%ebp
              jmp  2f
1:            shrl $2,%%edx
              movb %c8(%%edx),%%dl
              xorb -1(%3,%%edi),%%dl\n"
#if TARGET_CPU == 6
"             movzbl %%dl,%%ebp\n"
#else
"             movl %%edx,%%ebp\n"
#endif
"             xorb %b4,%%dl
              shrl $8,%4\n"
#if TARGET_CPU == 6
"             movzbl %%dl,%%edx\n"
#else
"             andl $0xff,%%edx\n"
#endif
"             xorl %c7(,%%edx,4),%4\n"
#if TARGET_CPU == 6
"             movzbl %b4,%%edx\n"
#else
"             movb %b4,%%dl\n"
#endif
"             addl %5,%%edx
              imul $134775813,%%edx
              leal 1(%%edx),%5
              shrl $24,%%edx\n"
#if TARGET_CPU == 6
"             xorl %6,%%edx
              shrl $8,%6
              movzbl %%dl,%%edx\n"
#else
"             xorb %b6,%%dl
              shrl $8,%6\n"
#endif
"             xorl %c7(,%%edx,4),%6
2:            "
#if TARGET_CPU == 6
"             movzwl %w6,%%edx\n"
#else
"             movw %w6,%%dx\n"
#endif
"             shrl $2,%%edx
              movb %c8(%%edx),%%dl
              xorb (%3,%%edi),%%dl\n"
#if TARGET_CPU == 6
"             movzbl %%dl,%%ebp\n"
#else
"             movl %%edx,%%ebp\n"
#endif
"             xorb %b4,%%dl
              shrl $8,%4\n"
#if TARGET_CPU == 6
"             movzbl %%dl,%%edx\n"
#else
"             andl $0xff,%%edx\n"
#endif
"             xorl %c7(,%%edx,4),%4\n"
#if TARGET_CPU == 6
"             movzbl %b4,%%edx\n"
#else
"             movb %b4,%%dl\n"
#endif
"             addl %5,%%edx
              imul $134775813,%%edx
              leal 1(%%edx),%5
              shrl $24,%%edx\n"
#if TARGET_CPU == 6
"             xorl %6,%%edx
              shrl $8,%6
              movzbl %%dl,%%edx\n"
#else
"             xorb %b6,%%dl
              shrl $8,%6\n"
#endif
"             xorl %c7(,%%edx,4),%6
              addl $2,%%edi\n"
#if TARGET_CPU == 6
"             movzwl %w6,%%edx\n"
#else
"             movw %w6,%%dx\n"
#endif
"             jne  1b\n"
#if TARGET_CPU == 6
"             movzwl %w6,%6\n"
#else
"             andl $0xffff,%6\n"
#endif
"             movb -1(%3),%1
              shrl $2,%6
              movl %%ebp,%k2
              popl %%ebp
              popl %%ecx
              xorb %c8(%6),%1
              popl %%ebx
              " : "=S" (b), "=a" (target), "=d" (pre_target)
                : "S" (b), "c" (key0), "a" (key1), "b" (key2),
                  "i" (crc_32_tab), "i" (mult_tab)
                : "edi"
            );
#         else
#           if !USE_MULT_TAB
              u16 t;
#           endif
            u32 kez0, kez1, kez2;
            u8 *e = b + FILE_SIZE - 1;
          
            kez0 = key0, kez1 = key1, kez2 = key2;
            do
              {
#               if USE_MULT_TAB
                  pre_target = *b++ ^ mult_tab [(u16)(kez2) >> 2];
#               else
                  t = kez2 | 2;
                  pre_target = *b++ ^ (u8)(((u16) (t * (t^1)) >> 8));
#               endif
                kez0 = crc32 (kez0, pre_target);
                kez1 = (kez1 + (u8)kez0) * 134775813 + 1;
                kez2 = crc32 (kez2, kez1 >> 24);
              }
            while (b < e);
            
#           if USE_MULT_TAB
              target = *b++ ^ mult_tab [(u16)(kez2) >> 2];
#           else
              t = kez2 | 2;
              target = *b++ ^ (u8)(((u16) (t * (t^1)) >> 8));
#           endif
#         endif

          /*printf ("pw=%s, t1=%02x, t2=%02x (%02x, %02x)\n", pw, target, pre_target, b[0], b[1]);*/
          
          if (target != *b++)
            goto out;
          
          if (pre_target == *b++)
            count2++;
        }
      while(--count);
      
      if ((changed = cbfunc (pw, 0)))
         return changed;
      
      out:
    }
  while ((changed = genfunc ()));
  
  return 0;
}



main.c
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#ifdef HAVE_STDBOOL_H
#include <stdbool.h>
#else
typedef enum { FALSE = 0, TRUE = 1 } bool;
#endif
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#else
#include "getopt.h"
#endif
#ifdef HAVE_GETTIMEOFDAY
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_SYS_TIME_H
#include <sys/time.h>
#endif
#endif

#include <string.h>

#include "crack.h"

static int use_unzip;
static method *crack_method = methods;
static int method_number = -1;
static int min_length = -1;
static int max_length = -1;
static int residuent = 0;
static int modul = 1;

static FILE *dict_file;

int REGPARAM
check_unzip (const char *pw)
{
  char buff[1024];
  int status;

#ifdef USE_UNIX_REDIRECTION
#define REDIR " >/dev/null 2>&1"
#else
#define REDIR ""
#endif

  sprintf (buff, "unzip -qqtP '%s' %s" REDIR, pw, file_path[0]);
  status = system (buff);

#undef REDIR

  if (status == EXIT_SUCCESS)
    {
      printf("\n\nPASSWORD FOUND!!!!: pw == %s\n", pw);
      exit (EXIT_SUCCESS);
    }

  return !status;
}

/* misc. callbacks.  */

static int
false_callback (const char *pw, const char *info)
{
  (void) pw;
  (void) info;                        /* suppress warning */
  return 0;
}

static int
true_callback (const char *pw, const char *info)
{
  (void) pw;
  (void) info;                        /* suppress warning */
  return 1;
}

static int
print_callback (const char *pw, const char *info)
{
  if (!use_unzip || check_unzip (pw))
    {
      printf ("possible pw found: %s (%s)\n", pw, info ? info : "");
      /*exit(0); */
    }

  return 0;
}

static int
brute_force_gen (void)
{
  u8 *p = pw_end;

  do
    {
      u8 o = *--p;
      *p = bf_next[o];
      if (o != bf_last)
        return pw_end - p;
    }
  while (p > pw);

  if (pw_end - pw < max_length)
    {
      p = ++pw_end;
      *p = 0;

      while (p > pw)
        *--p = bf_next[255];

      return -1;
    }
  else
    return 0;
}

static int
dictionary_gen (void)
{
  /* should optimize this, comparing prefixes would be a net win.
   * however, not using fgets but something better might be an
   * even higher win :(
   */
  if (fgets (pw, MAX_PW+1, dict_file))
    {
      pw[strlen (pw) - 1] = 0;
      return -1;
    }
  else
    {
      if (!feof (dict_file))
        perror ("dictionary_read_next_password");

      return 0;
    }
}

static int
validate_gen (void)
{
  return 0;
}

static void
validate (void)
{
  u8 header[HEADER_SIZE + 1] =
  {0xf4, 0x28, 0xd6, 0xee, 0xd7, 0xd2,
   0x3c, 0x1a, 0x20, 0xab, 0xdf, 0x73,
   0xd6, 0xba, 0};                /* PW: "Martha" */
  strcpy ((char *) files, (char *) header);        /* yeah, dirty... */
  file_count = 1;

  crack_method->init_crack_pw ();

  strcpy (pw, "Martha");
  if (crack_method->crack_pw (validate_gen, true_callback))
    printf ("validate ok (%s == Martha)\n", pw);
  else
    printf ("validation error (%s != Martha)\n", pw);
}

static void
parse_charset (char *cs)
{
  u8 chars[800];
  u8 map[256];
  u8 *p = chars;

  while (*cs)
    switch (*cs++)
      {
      case 'a':
        strcpy ((char *) p, "abcdefghijklmnopqrstuvwxyz");
        p += 26;
        break;

      case 'A':
        strcpy ((char *) p, "ABCDEFGHIJKLMNOPQRSTUVWXYZ");
        p += 26;
        break;

      case '1':
        strcpy ((char *) p, "0123456789");
        p += 10;
        break;

      case '!':
        strcpy ((char *) p, "!:$%&/()=?{[]}+*~#");
        p += 18;
        break;

      case ':':
        while (*cs)
          *p++ = *cs++;
        break;

      default:
        fprintf (stderr, "unknown charset specifier, only 'aA1!:' recognized\n");
        exit (1);
      }

  *p = 0;

  p = chars;
  bf_last = *p++;
  memset (bf_next, bf_last, sizeof bf_next);
  memset (map, 0, 256);

  for (; *p; p++)
    if (!map[*p])
      {
        map[*p] = 1;
        bf_next[bf_last] = *p;
        bf_last = *p;
      }

  bf_next[bf_last] = chars[0];

/*  { int i; for (i = 0; i < 255; i++) printf ("bf_next [%3d] = %3d\n", i, bf_next[i]);}; */
}

static int benchmark_count;

static int
benchmark_gen (void)
{
  if (!--benchmark_count)
    return 0;

  return brute_force_gen ();
}

static void
benchmark (void)
{
#ifdef HAVE_GETTIMEOFDAY
  int i;
  long j, k;
  struct timeval tv1, tv2;

  do
    {
      for (i = 0; i < HEADER_SIZE * 3; i++)
        files[i] = i ^ (i * 3);

      file_count = 3;
      strcpy (pw, "abcdefghij");
      parse_charset ("a");
      benchmark_count = BENCHMARK_LOOPS;

      verbosity = 0;

      printf ("%c%s: ",
              (crack_method - methods == default_method) ? '*' : ' ',
              crack_method->desc);

      if (strncmp ("zip", crack_method->desc, 3))
        printf ("(skipped)");
      else
        {
          fflush (stdout);

          crack_method->init_crack_pw ();
          gettimeofday (&tv1, 0);
          crack_method->crack_pw (benchmark_gen, false_callback);
          gettimeofday (&tv2, 0);
          tv2.tv_sec -= tv1.tv_sec;
          tv2.tv_usec -= tv1.tv_usec;

          j = tv2.tv_sec * 1000000 + tv2.tv_usec;
          k = BENCHMARK_LOOPS;

          printf ("cracks/s = ");

          for (i = 7; i--;)
            printf ("%ld", k / j), k = (k - k / j * j) * 10;
        }

      printf ("\n");
      crack_method++;
    }
  while (method_number < 0 && crack_method->desc);
#else
  fprintf (stderr, "This executable was compiled without support for benchmarking\n");
  exit (1);
#endif
}

static void
usage (int ec)
{
  fprintf (stderr, "\n"
           PACKAGE " version " VERSION ", a fast/free zip password cracker\n"
           "written by Marc Lehmann <pcg@goof.com> You can find more info on\n"
           "http://www.goof.com/pcg/marc/\n"
           "\n"
 	   "USAGE: fcrackzip\n"
           "          [-b|--brute-force]            use brute force algorithm\n"
           "          [-D|--dictionary]             use a dictionary\n"
           "          [-B|--benchmark]              execute a small benchmark\n"
           "          [-c|--charset characterset]   use characters from charset\n"
           "          [-h|--help]                   show this message\n"
           "          [-V|--validate]               sanity-check the algortihm\n"
           "          [-v|--verbose]                be more verbose\n"
           "          [-p|--init-password string]   use string as initial password/file\n"
           "          [-l|--length min-max]         check password with length min to max\n"
           "          [-u|--use-unzip]              use unzip to weed out wrong passwords\n"
           "          [-m|--method num]             use method number \"num\" (see below)\n"
           "          [-2|--modulo r/m]             only calculcate 1/m of the password\n"
           "          file...                    the zipfiles to crack\n"
           "\n"
    );

  fprintf (stderr, "methods compiled in (* = default):\n\n");
  for (crack_method = methods; crack_method->desc; crack_method++)
    fprintf (stderr, "%c%d: %s\n",
             (crack_method - methods == default_method) ? '*' : ' ',
             crack_method - methods,
             crack_method->desc);

  fprintf (stderr, "\n");
  exit (ec);
}

static struct option options[] =
{
  {"brute-force", no_argument, 0, 'b'},
  {"dictionary", no_argument, 0, 'D'},
  {"benchmark", no_argument, 0, 'B'},
  {"charset", required_argument, 0, 'c'},
  {"help", no_argument, 0, 'h'},
  {"validate", no_argument, 0, 'V'},
  {"verbose", no_argument, 0, 'v'},
  {"init-password", required_argument, 0, 'p'},
  {"length", required_argument, 0, 'l'},
  {"use-unzip", no_argument, 0, 'u'},
  {"method", required_argument, 0, 'm'},
  {"modulo", required_argument, 0, 2},
  {0, 0, 0, 0},
};

int
main (int argc, char *argv[])
{
  int c;
  int option_index = 0;
  char *charset = "aA1!";
  enum { m_benchmark, m_brute_force, m_dictionary } mode = m_brute_force;

  while ((c = getopt_long (argc, argv, "DbBc:hVvp:l:um:2:", options, &option_index)) != -1)
    switch (c)
      {
      case 'b':
        mode = m_brute_force;
        break;

      case 'D':
        mode = m_dictionary;
        break;

      case 'p':
        strcpy (pw, optarg);
        break;

      case 'l':
        pw[0] = 0;
        switch (sscanf (optarg, "%d-%d", &min_length, &max_length))
          {
          default:
            fprintf (stderr, "'%s' is an incorrect length specification\n", optarg);
            exit (1);
          case 1:
            max_length = min_length;
          case 2:
          }
        break;

      case 2:
        if (sscanf (optarg, "%d/%d", &residuent, &modul) != 2)
          fprintf (stderr, "malformed --modulo option, expected 'residuent/modul'\n"), exit (1);

        if (residuent < 0 || modul <= 0)
          fprintf (stderr, "residuent and modul must be positive\n"), exit (1);

        if (residuent >= modul)
          fprintf (stderr, "residuent must be less than modul\n"), exit (1);

        break;

      case 'B':
        mode = m_benchmark;
        benchmark ();
        exit (0);

      case 'v':
        verbosity++;
        break;

      case 'm':
        {
          for (method_number = 0; methods[method_number].desc; method_number++)
            if (!strncmp (methods[method_number].desc, optarg, strlen (optarg)))
              break;

          if (!methods[method_number].desc)
            method_number = atoi (optarg);

          crack_method = methods + method_number;
        }
        break;

      case 'V':
        validate ();
        exit (0);

      case 'c':
        charset = optarg;
        break;

      case 'u':
        use_unzip = 1;
        break;

      case 'h':
        usage (0);

      case ':':
        fprintf (stderr, "required argument missing\n");
        exit (1);

      case '?':
        fprintf (stderr, "unknown option\n");
        exit (1);

      default:
        usage (1);
      }

  if (method_number < 0)
    {
      method_number = default_method;
      crack_method = methods + default_method;
    }

  if (optind >= argc)
    {
      fprintf (stderr, "you have to specify one or more zip files (try --help)\n");
      exit (1);
    }

  for (; optind < argc; optind++)
    if (file_count < MAX_FILES)
      crack_method->load_file (argv[optind]);
    else if (verbosity)
      printf ("%d file maximum reached, ignoring '%s'\n", MAX_FILES, argv[optind]);

  if (file_count == 0)
    {
      fprintf (stderr, "no usable files found\n");
      exit (1);
    }

  crack_method->init_crack_pw ();

  switch (mode)
    {
    case m_brute_force:
      parse_charset (charset);

      if (!pw[0])
        {
          if (min_length < 0)
            {
              fprintf (stderr, "you have to specify either --init-password or --length with --brute-force\n");
              exit (1);
            }
          else
            {
              u8 *p = pw;
              while (p < pw + min_length)
                *p++ = bf_next[255];

              *p++ = 0;
            }
        }

      if (residuent)
        {
          int xmodul = modul;
          modul = residuent;
          pw_end = pw + strlen (pw);
          brute_force_gen ();
          printf ("%s\n", pw);
          modul = xmodul;
          printf ("WARNING: residuent mode NOT supported YET!\n");
        }

      crack_method->crack_pw (brute_force_gen, print_callback);
      break;

    case m_dictionary:
      if (!pw[0])
        {
          fprintf (stderr, "you have to specify a file to read passwords from using the  -p switch\n");
          exit (1);
        }

      if (!(dict_file = fopen (pw, "r")))
        {
          perror (pw);
          exit (1);
        }
      else
        {
          *(pw_end = pw) = 0;
          dictionary_gen (); /* fetch first password */
          crack_method->crack_pw (dictionary_gen, print_callback);

          fclose (dict_file);
        }

      break;

    default:
      fprintf (stderr, "specified mode not supported in this version\n");
      exit (1);
    }

  return 0;
}
